### ISSUE TITLE
Optimize batched RPC lookups in P2P branch

### ISSUE THREAD

**Issue by Pantamis** [romanz/electrs#464]
Batched RPC can be resolved faster in P2P branch.

Indeed, each RPC will require loading one block: the one at the height given by an index lookup. But several requests may require looking at the same block ! Currently, each RPC will load the block needed once no matter what.

An interesting improvement could be to first get all needed block heights for each RPC, sort the requests by block height and load a block only once even if it is required for several RPCs. Once loaded we search all the transactions we are looking for in it and send the responses in the order at which we find them.

This is very similar to what is done when loading blocks with scripthash status sync where several blocks are loaded but one block may contain several relevant transactions.

However I think this may require significant refactoring of the code....

**Comment by Pantamis** [comment]
I tested 70025e4 with my watch only JoinMarket wallet (thousands of subscriptions) in sparrow, the difference is very significant ! The UX is so much better ! (even better than ElectrumX !)

I don't think the impact is caused by smarter blocks lookups but rather parallel subscriptions because the scripthashes rarely appear in the same block in my wallet.

I guess this is not a priority currently but it would be really nice to have parallel subscriptions with smart blocks lookups for outpoints subscriptions for personal block explorers, something to add in #454. Smart block lookup would be very powerful to load all spending transactions of outpoints of a big transaction in janoside/btc-rpc-explorer#356 because many such outpoints are often spend at the same block (in the 10 next blocks). Not having to re-parse the same block would have a great impact on the loading time.

**Comment by Kixunil** [comment]
@Pantamis that needs a spec change but I don't see why it wouldn't be made - it is very reasonable and simple.
